<!DOCTYPE html>
<html lang="en-US" dir="ltr">
  <head>
    <base href="/mount/ultralaser_home/Projects/c2_wiki_recovery/out/">
    <meta charset="utf-8">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <link rel="stylesheet" href="style.css">
  </head>
  <body>
    <div class="page">
      <h1>
        <img src="wiki.gif" />
        Java Unit Test Challenge Solved
      </h1>
      <p>
        As part of <a href="ExtremeProgrammingChallengeFourteen.html">ExtremeProgrammingChallengeFourteen</a>, <a href="TomCargill.html">TomCargill</a> gave us some Java code and asked us <a href="ExtremeProgrammer.html">ExtremeProgrammer</a>s to test it and show that it has a bug of some sort. This is what he gave us:
      </p>
      <code>
        import java.lang.*;<br/>
        public class <a href="BoundedBuffer.html">BoundedBuffer</a> extends java.lang.Object {<br/>
        public Object[] buffer = new Object[4];<br/>
        public int putAt, takeAt, occupied;<br/>
      </code>
      <code>
        synchronized void put(Object x) throws <a href="InterruptedException.html">InterruptedException</a> {<br/>
        while(occupied == buffer.length)wait();<br/>
        notify();<br/>
        ++occupied;<br/>
        putAt %= buffer.length;<br/>
        buffer[putAt++] = x;}<br/>
      </code>
      <code>
        synchronized Object take() throws <a href="InterruptedException.html">InterruptedException</a> {<br/>
        while(occupied == 0)wait();<br/>
        notify();<br/>
        --occupied;<br/>
        takeAt %= buffer.length;<br/>
        return buffer[takeAt++];}<br/>
        }<br/>
      </code>
      <p>
        I took on the task of testing it myself to see what could be done with it. Can we show that it has a bug? Let's see. -- <a href="DonWells.html">DonWells</a>
      </p>
      <hr/>
      <p>
        The first thing we need is of course a <a href="SimpleJavaUnitTestFramework.html">SimpleJavaUnitTestFramework</a>.  Now that we have that we can add some simple tests, maybe we will see something wrong. I am going to <a href="TestThePutMethodFirst.html">TestThePutMethodFirst</a> in a stand alone fashion. Then I <a href="TestTheTakeMethodSecond.html">TestTheTakeMethodSecond</a> in a stand alone fashion. No bug yet. I was wrong about there being subtle bugs in the boundary conditions.
      </p>
      <p>
        Now I have to <a href="ExtendTheJavaUnitTestingFramework.html">ExtendTheJavaUnitTestingFramework</a>. The <a href="TestingFramework.html">TestingFramework</a> now detects when a deadlock occurs by only letting a test run for a maximum of 5 seconds. Now I feel safe about using real threads in my tests.  
      </p>
      <p>
        I can add a <a href="TestForPutInAthread.html">TestForPutInAthread</a>. I can add a <a href="TestForTakeInAthread.html">TestForTakeInAthread</a> too. Now I have the building blocks to <a href="TestBothThreadsRunning.html">TestBothThreadsRunning</a> simultaneously. That works too. And next I want to <a href="TestFourThreadsRunning.html">TestFourThreadsRunning</a> simultaneously. This works too.  I thought it would show a bug, but again <a href="UnitTest.html">UnitTest</a>s often show you things are actually working. If this were part of an actual project I would create a few more tests. The first would be to <a href="TestTwentyThreadsRunning.html">TestTwentyThreadsRunning</a> simultaneously.  After that I add a test which tests say <a href="TenPutsAndOneTake.html">TenPutsAndOneTake</a>. Then of course, the opposite, one put and ten takes. And now <a href="AbugIsFound.html">AbugIsFound</a>. I was thinking though that we should have found this earlier. I decide to go back and <a href="AddOneMoreCheckToTestTwentyThreads.html">AddOneMoreCheckToTestTwentyThreads</a>. Now I am going to remove three of the tests. One is too intrusive and two are redundant. Now as I run this code and listen to it I find that it does not run the same all the time. It is non-deterministic because there is some randomness built into the Java runtime. I need to <a href="ExtendTheJavaTestFrameworkForNondeterminism.html">ExtendTheJavaTestFrameworkForNondeterminism</a>. Now we can see there is indeed a bug.  
      </p>
      <p>
        <img src="http://members.aol.com/jdwells123/twofailed.gif" />
      </p>
      <p>
        Fortunately, Tom gave us a quick fix when he gave us <a href="ExtremeProgrammingChallengeFourteenTheBug.html">ExtremeProgrammingChallengeFourteenTheBug</a>. We can simply change 
      </p>
      <code>
        notify();<br/>
      </code>
      <p>
        into
      </p>
      <code>
        notifyAll();<br/>
      </code>
      <p>
        in both the take and put messages. Let's run our tests again:
      </p>
      <p>
        <img src="http://members.aol.com/jdwells123/allpass.gif" />
      </p>
      <p>
        We are done. -- <a href="DonWells.html">DonWells</a>
      </p>
      <hr/>
      <p>
        Here is the source for the latest version in zip format.
      </p>
      <p>
        <a href="ftp://members.aol.com/jdwells123/challenge14.zip">ftp://members.aol.com/jdwells123/challenge14.zip</a>
      </p>
      <p>
        I believe it is ok now.
      </p>
      <hr/>
      <p>
        Not that I'm sure anybody cares anymore, but why was everyone so hung up about the non-determinism?
      </p>
      <code>
        public static void testConcurrency() throws <a href="InterruptedException.html">InterruptedException</a> {<br/>
        final int[] threadsFinished = new int[]{0};<br/>
        final <a href="BoundedBuffer.html">BoundedBuffer</a> buffer = new <a href="BoundedBuffer.html">BoundedBuffer</a>();<br/>
      </code>
      <code>
        Thread thread1 = new Thread() {<br/>
        public void run() {<br/>
        synchronized (buffer) {<br/>
        synchronized (this) {<br/>
        this.notifyAll();<br/>
        }<br/>
        try {<br/>
        buffer.wait();								<br/>
        } catch (<a href="InterruptedException.html">InterruptedException</a> e) {}<br/>
        threadsFinished[0]++;<br/>
        }<br/>
        }<br/>
        };<br/>
        Thread thread2 = new Thread() {<br/>
        public void run() {<br/>
        synchronized (buffer) {<br/>
        synchronized (this) {<br/>
        this.notifyAll();<br/>
        }<br/>
        try {<br/>
        buffer.wait(); 										<br/>
        } catch (<a href="InterruptedException.html">InterruptedException</a> e) {}<br/>
        threadsFinished[0]++;<br/>
        }<br/>
        }<br/>
        };<br/>
      </code>
      <code>
        synchronized (thread1) {<br/>
        thread1.start();<br/>
        thread1.wait();<br/>
        }<br/>
        synchronized (thread2) {<br/>
        thread2.start();<br/>
        thread2.wait();<br/>
        }<br/>
        synchronized (buffer) {<br/>
        buffer.put("");<br/>
        thread1.join(1000);<br/>
        thread2.join(1000);<br/>
        }<br/>
        assertEquals("Not all test threads were awoken!", 2, threadsFinished[0]);<br/>
        }<br/>
      </code>
      <p>
        Worked the first time, by which I mean, caught the error on the first (and all subsequent runs), and passed on all runs after the notify was changed to notifyAll().  If anything, this would give a false positive, which in the case of concurrency, is much preferable to a false negative one time in a thousand.  
      </p>
      <p>
        The trick in this case is that a usage of 'notify()' vs. 'notifyAll()' can only ever wake up a single thread, and therefore can never wake up both test threads.
      </p>
      <p>
        Now, a good case could be made for the point that the synchronization object <em>should</em> be an implementation detail;  it should really be an internal object being locked on, not the public instance.  This would render this particular test useless.  So we design for testability, passing in the object to be locked on in the constructor, while still providing a constructor to provide a unique object by default.
      </p>
      <p>
        --<a href="WilliamUnderwood.html">WilliamUnderwood</a>
      </p>
      <p>
        I haven't examined this page in great detail, but your question "why was everyone so hung up about the non-determinism?" -- you know, the answer to <strong>that</strong> question never depends on the topic at hand. The answer is, "how can you prove that the non-determinism will <strong>never</strong> bite you?<em></em>
      </p>
      <p>
        In other words, the real question is not whether a test catches a known non-determinism problem, the question is whether the test is absolutely <strong>guaranteed</strong> to -- which you didn't comment on. (I trust you are aware that no amount of testing, in itself, can answer that question, since you might get lucky on each test.) -- dm
      </p>
      <dl>
        <dt> </dt>
        <dd>My point was actually that the first statement of the challenge was to detect a known condition.  The 'so hung up about non-determinism' was more of a commentary of the solutions that I saw;  basically, I feel they all erred on the wrong side of the coin:  why create a test which is known to give false negatives, and that you can therefore never be sure of?  Testing has enough holes already, let's not add more.  </dd>
      </dl>
      <dl>
        <dt> </dt>
        <dd>My improved test shows a single property:  that the implementation of 'put()' does not use 'notify()'.  That's sufficient if the 'while(<expression>) {wait()}' is used.  It would always be sufficient to make sure that 'do{ wait(); notify(); }while(<expression>);' is used.</dd>
      </dl>
      <dl>
        <dt> </dt>
        <dd>Basically, I _<strong>really</strong>_ don't like false negatives in my tests.  They're inevitable, but it's best to weed out as many as possible.</dd>
      </dl>
      <dl>
        <dt> </dt>
        <dd>--cwillu</dd>
      </dl>
    </div>
  </body>
</html>